compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -3425,9 +3441,13 @@ class Typer extends Namer
       tree match {
         case closure(Nil, id @ Ident(nme.ANON_FUN), _)
         if defn.isFunctionType(wtp) && !defn.isFunctionType(pt) =>
-          pt match {
+          // Strip the Null type from the pt before check SAM type
+          val pt1 = if ctx.explicitNulls then pt.stripNull else pt
```

Comment:
This expression appears 6 times, so it's worth factoring this out into a separate method. Or maybe we can make `stripNull` return the type itself if `ctx.explicitNulls` is not set? It's worth experimenting here!

Reply:
If we define `stripNull` in this way, I'm worry this may cause confusion to other developers when looking at its name. How about renaming this function to `stripNullWhenExplicit`?